### DISCUSSION TITLE
Embedded languages - all features

### DISCUSSION THREAD

**Discussion by AllanJard** [microsoft/vscode-discussions#606] [Extension Development QnA] [answered]
I'm trying to figure out how I have have Javascript code completion, formatting, etc, inside an XML file. Based on the [examples](https://code.visualstudio.com/api/language-extensions/embedded-languages) I can see how to do auto complete, and I can do syntax highlighting with a suitable grammer. But I can't see how to do formatting.

Is there a middleware for that (I presume that is where it goes)? Or an example?

Or is there a way to just say "this part of the document is CSS - offer all the CSS stuff". At the moment it looks like I might need a bunch of different middlewares for the different editor operations?

**Edit** I suppose, another way of putting this is, how would I add formatting support to the LSP embedded examples?

**Comment by r3m0t** [comment]
It's analogous, you define `provideDocumentFormattingEdits` instead of `provideCompletionItem` and you call `vscode.executeFormatDocumentProvider` instead of `vscode.executeCompletionItemProvider`.

The code would be more complicated as you would want to format the XML and CSS separately then combine them together.

The list of commands is here- https://code.visualstudio.com/api/references/commands

If you use Typescript you will get autocompletion on the method names you can put in a middleware, otherwise they are here- https://github.com/microsoft/vscode-languageserver-node/blob/1320922f95ef182df2cf76b7c96b1a2d3ba14c2a/client/src/common/client.ts#L347 / https://github.com/microsoft/vscode-languageserver-node/blob/1320922f95ef182df2cf76b7c96b1a2d3ba14c2a/client/src/common/formatting.ts#L44

I don't think anybody has made a reusable request forwarding example, with all 20+ middleware.

**Comment by AllanJard** [comment]
Many thanks.

If I add this:

```js
			provideDocumentFormattingEdits: async ( document, options, token, next) => {
				console.log('FORMAT', document);
				return await next(document, options, token);
			}
```

to the middleware for the example (which creates an `html1` language with CSS embedded in a `style` tag), then I just get a message from VS Code saying "There is no formatter for 'html1' files installed".

I tried `provideDocumentRangeFormattingEdits` as well, but there is no formatting options shown in the context menu if I select the CSS and right click.

My goal at the moment is simply to pick out the CSS (which I can do) and format that - I'm not bothered about the surrounding HTML at this time.

Do I need to create a second plug-in that would be a formatter for the file? If so, I don't understand what `provideDocumentFormattingEdits` does in the language client?

**Reply by r3m0t** [reply, marked as answer]
Hi, the middleware needs to follow this example:

https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding-sample

Ie a new uri with new contents just containing the css part of the xml document.

Then the response line and character positions need to be modified to correspond to the xml document.
